lib/webdriveragent.js

```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
-        if (e.length && d.indexOf(AGENT_LOG_PREFIX) > -1) {
-          agentLog.error(`Sim: ${e}`);
-        }
-      });
-      this.simLogs.on('exit', (code) => {
-        reject(new Error(`tailing of simulator log exited with code ${code}`));
-        return;
-      });
-
-      this.xcodebuild = this.createXcodeBuildSubProcess();
-      this.xcodebuild.on('output', (d, e) => {
-        // we want to pull out the log file that is created, and highlight it
-        // for diagnostic purposes
-        if (e.indexOf('Writing diagnostic log for test session to') !== -1) {
-          // pull out the first line that begins with the path separator
-          // which should be the line indicating the log file generated
-          this.logLocation = _.first(_.remove(e.trim().split('\n'), (v) => v.indexOf(path.sep) === 0));
-          log.debug(`Log file for xcodebuild test: ${this.logLocation}`);
-        }
-        // log output of xcodebuild, for debugging
-        // log.info(d, e);
-      });
-      this.xcodebuild.on('exit', (code, signal) => {
-        if (!code && signal === 'SIGTERM') {
-          log.info(`xcodebuild exited with code '${code}' and signal '${signal}'`);
-        }
-        if (!signal && code !== 0) {
-          reject(new Error(`xcodebuild failed with code ${code}`));
-          return;
-        }
-        this.quit();
-      });
+      try {
+        this.deviceLogs = realDevice ? this.createRealDeviceLogsSubProcess()
+                                     : this.createSimLogsSubProcess();
+      } catch (err) {
+        reject(err);
+      }
 
-      await this.checkForDependencies();
+      try {
+        this.xcodebuild = this.createXcodeBuildSubProcess(reject, realDevice);
 
-      await this.xcodebuild.start();
+        await this.checkForDependencies();
 
-      // we have to wait for the sim to start before we can tail the log file
-      await systemLogExists(this.sim);
+        let startTime = Date.now();
+        await this.xcodebuild.start();
 
-      let agentUrl;
-      let agentStartedOnDevice = this.simLogs.start((stdout) => {
-        let match = AGENT_STARTED_REGEX.exec(stdout);
-        if (match) {
-          agentUrl = match[1];
-          log.info(`detected that WebDriverAgent is running at url ${agentUrl}`);
-          if (!agentUrl) {
-            log.errorAndThrow(new Error('No url detected from WebDriverAgent'));
-          }
-          return true;
+        if (!realDevice) {
+          // we have to wait for the sim to start before we can tail the log file
+          await systemLogExists(this.device);
         }
-      });
 
-      log.info(`Waiting for WebDriverAgent to start on device`);
-      await agentStartedOnDevice;
-      log.info(`WebDriverAgent started at url ${agentUrl}`);
+        let agentUrl;
+        let lineCount = 0;
+        let reachedEnd = realDevice ? false : true; // simulator does not need to wait
+        let agentStartedOnDevice = this.deviceLogs.start((stdout) => {
```

Comment:
rather than awaiting the bare promise below, this might look a little better like:

```js
const startDetector = (stdout) => {
  ...
}

await this.deviceLogs.start(startDetector);
```